enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+    For details see the `scaleUp` configuration.
+    An empty list of triggers disables scale down.
+
+A complete custom resources instance in YAML format could then look like this:
+
+```yaml
+controlplaneautoscaling:
+  machineConfiguration:
+    machineType: machines_v1beta1_machine_openshift_io
+    machines_v1beta1_machine_openshift_io:
+      machineConfigurations:
+        platform: AWS
+        aws:
+        - weight: 1
+          instanceSize: r4.large
+        - weight: 2
+          instanceSize: r4.xlarge
+  syncPeriod: [duration as specified in golang: e.g. 5s, 15s]
+  scaleUp:
+    selectPolicy: "next"
+    stabilizationWindow: [duration as specified in golang: e.g. 5s, 15s]
+    triggerPolicy: "any"
+    triggers:
+      - type: "cpu"
+        value: "80"
+        timeWindow: "30m"
+  scaleDown:
+    selectPolicy: "next"
+    stabilizationWindow: [duration as specified in golang: e.g. 5s, 15s]
+    triggerPolicy: "all"
+    triggers:
+      - type: "cpu"
+        value: "80"
+        timeWindow: "30m"
+      - type: "memory"
+        value: "90"
+        timeWindow: "30m"
+```
+
+In case the user decides to use `prometheus`, authentication will be specified
+in an additional custom resource
+`ControlPlaneMachineSetAutoscalingTriggerAuthentication` that is based on
+[CMA](https://docs.openshift.com/container-platform/4.13/nodes/cma/nodes-cma-autoscaling-custom-trigger.html#nodes-cma-autoscaling-custom-prometheus-config_nodes-cma-autoscaling-custom-trigger
+) as well and specifies the following values:
+
+- `secretTargetRef`:
+  - `parameter`: type of the secret referenced: should be `bearer` for bearer
+    authentication.
+  - `name`: name of the secret too use.
+  - `key`: key in the secret that contains the token.
+
+### Data gathering
+
+As specified the operator should be able to use two different data sources:
+metrics API and prometheus.
+
+#### Metrics API
+
+The operator can gather the following data via the [metrics
+API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/):
+
+- CPU usage in `cores`.
+- Memory usage in a SI units.
+- The time window used to aggregate the above mentioned data.
+
+To gather data about the available CPU cores and memory available to control
+plane nodes, the operator should use the kubernetes
+[Node](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#node-v1-core)
+API.
+
+As the window returned by metrics API is likely shorter than required window for
+scaling decisions, the operator should gather its own average to allow making
+the scaling decision based on the metrics data.
+
+#### Prometheus
+
+Prometheus queries should be configured to use the correct window already.
+
+The operator has to handle a metric's data missing or authentication failure.
+
+Both cases should use a conservative approach of not performing any actions, and
+not interrupting cluster operation.
```

Comment:
Will you expect this to cause the operator to mark itself as degraded? How as an end user would I know that the scaling is not working as I expect it to